src/coap_io.c

```diff
@@ -1424,6 +1518,69 @@ coap_run_once(coap_context_t *ctx, unsigned timeout_ms) {
   coap_ticks(&now);
   coap_read(ctx, now);
 
+#else /* HAVE_EPOLL_SUPPORT */
+  do {
+    struct epoll_event events[COAP_MAX_EPOLL_EVENTS];
+    int etimeout = timeout;
+
+    /* Potentially adjust based on what the caller wants */
+    if (timeout_ms == 0)
+      etimeout = -1;
+    else if (timeout_ms == 1)
+      etimeout = 0;
+
+    nfds = epoll_wait(ctx->epfd, events, COAP_MAX_EPOLL_EVENTS, etimeout);
+    if (nfds < 0) {
+      if (errno != EINTR) {
+        coap_log (LOG_ERR, "epoll_wait: unexpected error: %s (%d)\n",
+                            coap_socket_strerror(), nfds);
+      }
+      break;
+    }
+
+    if (coap_io_do_events(ctx, events, nfds)) {
+      /* timeout event occurred, need to see what timedout for processing */
+      coap_ticks(&now);
+      coap_write(ctx, sockets,
+                 (unsigned int)(sizeof(sockets) / sizeof(sockets[0])),
+                  &num_sockets, now);
+      /* fake nfds so we go around the loop again */
+      nfds = COAP_MAX_EPOLL_EVENTS;
+    }
+
+    /*
+     * reset to 1 (which causes etimeout to become 0)
+     * incase we have to do another iteration
+     */
+    timeout_ms = 1;
+    /* Keep retrying until less than COAP_MAX_EPOLL_EVENTS are returned */
+  } while (nfds == COAP_MAX_EPOLL_EVENTS);
+
+  if (ctx->eptimerfd != -1) {
+    struct itimerspec new_value;
+    struct itimerspec old_value;
+    int ret;
+
+    memset(&new_value, 0, sizeof(new_value));
+    coap_ticks(&now);
+    if (ctx->next_timeout != 0 && ctx->next_timeout > now) {
+      coap_tick_t rem_timeout = ctx->next_timeout - now;
+      /* Need to trigger an event on ctx->epfd in the future */
+      new_value.it_value.tv_sec = rem_timeout / 1000;
```

Comment:
To calculate seconds from coap ticks, you need to divide by `COAP_TICKS_PER_SECOND` which may or may not have the value 1000.

Reply:
Corrected in the subsequent line as well.